packages/legend-application-studio/src/stores/EditorStore.ts

```diff
@@ -1187,11 +922,77 @@ export class EditorStore implements CommandRegistrar {
     );
   }
 
-  changeLocalChangesState(): void {
-    if (this.isInFormMode) {
-      this.localChangesState = new FormLocalChangesState(this, this.sdlcState);
-    } else {
-      this.localChangesState = new TextLocalChangesState(this, this.sdlcState);
+  *switchModes(
+    to: GRAPH_EDITOR_MODE,
+    fallbackOptions?: {
+      isCompilationFailure?: boolean;
+      isGraphBuildFailure?: boolean;
+    },
+  ): GeneratorFn<void> {
+    switch (to) {
+      case GRAPH_EDITOR_MODE.GRAMMAR_TEXT: {
+        const graphEditorMode = new GraphEditSingleFileGrammarModeState(this);
+        try {
+          yield flowResult(
+            graphEditorMode.cleanupBeforeEntering(fallbackOptions),
+          );
+          this.graphEditorMode = graphEditorMode;
+          yield flowResult(
+            this.graphEditorMode.initialize(Boolean(fallbackOptions)),
+          );
+        } catch (error) {
+          assertErrorThrown(error);
+          this.applicationStore.notifyWarning(
+            `Can't enter text mode: transformation to grammar text failed. Error: ${error.message}`,
+          );
+          this.applicationStore.setBlockingAlert(undefined);
+          return;
+        }
+        break;
+      }
+      case GRAPH_EDITOR_MODE.FORM: {
+        if (this.graphState.checkIfApplicationUpdateOperationIsRunning()) {
+          return;
+        }
+        try {
+          try {
+            yield flowResult(
+              this.getGraphEditorMode(
```

Comment:
something here is still not right to me, I want the treatement here to be very simple and symmetrical. Say every time one changes the mode, we do:

```ts
const newMode = new ...();

// is this needed?
try {
  // prepare new state
  newMode.prepare(); // prepare the material needed to initialize() new state, or the necessary checks, if this fail, we do not allow leaving current mode since new state is not possible
} catch (err) {
  ... // if failure here, we probably should abort leaving mode?
}

try {
  currentMode.leave() // clean up --> this implies we need to set some state somewhere before starting a new mode, we should really minimize the logic here and offload to `initialize` as much as possible, also, there should not be failure happening
} catch (err) {
  ... // handle and potentially disallow leaving the mode
}

try {
  newMode.initialize() // enter new mode (bulk of the logic is here)
} catch (err) {
  ... // handle here, if failed to initialize new mode, we need a fall back mechanism to handle
  // maybe do this?
  newMode.fallback();
}
```

I think if we break things down like that, it would be much easier.